### PR TITLE
AB#750: Add initial support for Occlum

### PR DESCRIPTION
**Pull request by Nirusu**
This PR adds:

- OE-compatible DCAP quote support for Occlum-based applications
- An Occlum premain process (`premain-occlum`), which is expected to be built using `occlum-go` (A Dockerfile is provided, too)
- `occlum-hello` sample, based on the premain and Occlum Docker image (+ a slightly modified version for Azure DCAP)
- CLI command `package-info`, which allows printing the properties needed for `manifest.json` for either Occlum or a generic SGX enclave application (though not sure how this is useful without a respective premain, but it's there)

Also, slightly out of scope:
- Adds copyright headers for libOS-related (premain) files

Note that the commands used in the README.md are not 100% copy-pasteable until we merged this, as they expect everything to be on the main branch. Also, later on we could replace the `occlum-go` build process with a download of a prebuilt binary in the Makefile, similar to handle this with Graphene.

Feedback & Testing appreciated!

### CODE REVIEW

File: cli/cmd/packageInfo.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+	// Get MRENCLAVE (UniqueID), ISVPRODID (ProductID) and ISVSVN (SecurityVersion) directly from SIGSTRUCT
+	// Get Modulus so we can calculate MRSIGNER (= SHA256 hash of modulus)
+	modulus := sigStruct[128:512]
+	mrenclave := sigStruct[960:992]
+	isvprodid := sigStruct[1024:1026]
+	isvsvn := sigStruct[1026:1028]
+
+	// Calculate MRSIGNER, which is the SHA-256 hash of the modulus stored in SIGSTRUCT
+	mrsigner := sha256.Sum256(modulus)
```
Comment (thomasten) [inline cli/cmd/packageInfo.go:109]: Can you move this block in a function and add a unit test for it?

**Review by thomasten — CHANGES_REQUESTED** [review]

File: cmd/premain-occlum/main.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+	argv := toCArray(os.Args)
+	envp := toCArray(os.Environ())
+
+	// Occlum cannot handle nil for the PID parameter ("pointer not in user space")
+	spawnedPID := C.int(0)
+
+	fmt.Printf("Exiting PreMain. Launching: %s\n", service)
+	// spawn service
+	if res := C.posix_spawn(&spawnedPID, C.CString(service), nil, nil, &argv[0], &envp[0]); res == -1 {
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply cmd/premain-occlum/main.go:61]: Returns -1 if fails to spawn the application, yes. The error messages listed down are actually errors I encountered (& fixed) myself when receiving -1 as error code.

File: cmd/premain-occlum/main.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		color.Red("ERROR: Failed to spawn the target process.")
+		color.Red("Did you specify the correct path for your target application in the Marblerun manifest as first parameter for 'Argv' (for example: /bin/hello_world)?")
+		color.Red("Have you allocated enough memory?")
+		panic(syscall.Errno(res))
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply cmd/premain-occlum/main.go:65]: Printing Errno()? Or the above errors? Cause I actually encountered them with -1.

File: samples/occlum-hello/Occlum.json [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    "entry_points": [
+        "/bin/premain-occlum"
+    ],
+    "env": {
+        "default": [
+            "OCCLUM=yes",
+            "EDG_MARBLE_COORDINATOR_ADDR=localhost:2001",
+            "EDG_MARBLE_TYPE=hello",
+            "EDG_MARBLE_UUID_FILE=uuid",
+            "EDG_MARBLE_DNS_NAMES=localhost"
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply samples/occlum-hello/Occlum.json:22]: Mhh, I don't really know. For the sake of the sample, I think it's nice to have them in here as default.

But we can also define them _additionally_ as untrusted if someone wants to override them. That should be supported according to the documentation.

File: cli/cmd/packageInfo.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	"os"
+	"path/filepath"
+
+	"github.com/fatih/color"
+	"github.com/spf13/cobra"
+)
+
+func newPackageInfoCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "package-info",
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply cli/cmd/packageInfo.go:20]: Nope, does not for those.

File: cli/cmd/packageInfo.go (hunk 1 above)
Comment (Nirusu) [reply cli/cmd/packageInfo.go:109]: Think this is similar to the situation we had with EGo where a test ELF file would be good or needed, however shipping one might bloat the unit tests a little bit.

If hardcoding / supplying a precompiled valid section is fine, that's no problem.

File: cmd/premain-occlum/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (thomasten) [reply cmd/premain-occlum/main.go:65]: just the syscall.Errno(-1) part

File: samples/occlum-hello/Occlum.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomasten) [reply samples/occlum-hello/Occlum.json:22]: > we can also define them _additionally_ as untrusted

I like

File: cli/cmd/packageInfo.go (hunk 1 above)
Comment (thomasten) [reply cli/cmd/packageInfo.go:109]: Yes, just add a valid section as hardcoded []byte slice to the test.

File: marble/premain/occlum.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	dcapCQuoteSize := C.uint32_t(sgxQuoteSize)
+	dcapQuoteBuffer := make([]byte, sgxQuoteSize)
+
+	// Generate reportData: SHA-256 hash of input data
+	hash := sha256.Sum256(cert)
+	reportData := make([]byte, 64)
+	bytesCopied := copy(reportData, hash[:])
+	if bytesCopied != 32 {
+		return nil, errors.New("too few bytes copied into report data for quote generation")
+	}
```
Comment (thomasten) [inline marble/premain/occlum.go:92]: you may do
```suggestion
	reportData := append(hash[:], make([]byte, 32)...)
```
but you can also keep as is if you think the final size is too implicit

**Review by thomasten — CHANGES_REQUESTED** [review]

File: marble/premain/occlum.go (hunk 6 above)
Comment (Nirusu) [reply marble/premain/occlum.go:92]: I think I would just keep it explicit, personally I like it more in terms of readability.

**Review by thomasten — APPROVED** [review]

**Pull request merged**